### ISSUE TITLE
Add multi-file-upload component

### ISSUE THREAD

**Issue by avishkar58** [deep-learning-indaba/Baobab#734]
Add a new component called `FormMultiFile` to webapp/components/form which allows a user to upload multiple files and set a name for each file uploaded. The "value" of this editor should be a list of objects containing a "name" and "id" for each file. 

Default the name to the file name.

Render this component in the application form for questions of type "multi-file"

**Comment by avishkar58** [comment]
More details about user flow:
- The control maintains a list of files
- A user can upload a new file by clicking on an "add file" button with a plus icon
- The name field is defaulted to the name of the file, but the user can edit this
- The user can update the name in any row in the list
- The user can open the file from any row by clicking on a "view file" link, which opens the file in a new tab
- The user can remove a file by clicking on a trash icon next to the row. Since the backend API doesn't support file deletion yet, this action can just remove the file from the list
- If the user saves the form and comes back to it later, they should still be able to view the files and change the names of the files they previously uploaded

**Comment by greeffjean** [comment]
Hi Avishkar.

A couple of questions:
In regards to editing files, back and forth I think this would be the best
way:
1. Create a file list.
2. On Form Submit upload all files at once.
3. On component re-render and edit file, upload new changes or simply
replace the whole list.

This logic above will require me to use API calls that post and edit a n
array or object of file data, do you agree with this logic?
Is there some examples of a API setup and logic I can see to help me with
this flow, at the moment I have nothing much to work with? Only a Upload
file function in ApplicationFrom.js

**Comment by greeffjean** [comment]
Thus far I have this component and children 80% built out the missing
pieces are just these steps:

   - The user can open the file from any row by clicking on a "view file"
   link, which opens the file in a new tab
   - If the user saves the form and comes back to it later, they should
   still be able to view the files and change the names of the files they
   previously uploaded

I am currently handling any updates to the list in the new components
state, but would you want me to post these updates to the API ? I ask
because there does not seem much for me to work with in terms of API
functionality examples and Im wondering what you meant by editing the list,
from where? from the API storage or the components local state?

**Comment by greeffjean** [comment]
Hi Avishkar.

Im not quite sure how to proceed with this task, I need some feedback from
you here.

How are you expecting me to edit the file names, in the local state or in
the API storage, currently no names are being stored in the API

**Comment by avishkar58** [comment]
Hi Jean,

I don't think you need a new API for this - as long as you keep your local state synchronised to the applicationForm's `answer.value` that will provide your saving and loading functionality and will keep things generic. Have a look at [FormMultiCheckbox](https://github.com/deep-learning-indaba/Baobab/blob/develop/webapp/src/components/form/FormMultiCheckbox.js) which does something similar to this (it initialises the component state to `props.defaultValue` and every time the local state changes it calls `props.onChange` to push that to the parent `answer.value` in ApplicationForm.js, essentially encoding the local state in a string. In your case, you can serialize a JSON representation of your state to store in `answer.value`. 

As another example, you can also look at [ReferenceRequest](https://github.com/deep-learning-indaba/Baobab/blob/develop/webapp/src/pages/applicationForm/components/ReferenceRequest.js) which also has a list of items, but this one does talk to a separate API because it also needs to kick off additional processes and not just save state.

**Comment by greeffjean** [comment]
Hi Avishkar.

Ive been a bit puzzled at times with the setup but I managed in the end, I
will volunteer to put extra hours in over the weekend to crack this issue
as I really wanted to complete this task by today, Im tied up with other
commitments from Wed - Friday but its also likely that I can put some
additional hours in the week as well.

I built out 90% of this component it's looking good and working well, the
missing piece of the puzzle is just keeping the state in sync with
answer.value of the Form.
Im needing to re-structure my component renders and add in some extra
functionality, watch this space.

Regards
Jean
